### DISCUSSION TITLE
RequestData type hint is dict, possibly should be Union[dict,str,bytes]

### DISCUSSION THREAD

**Discussion by arthur-tacca** [encode/httpx#2146] [Potential Issue]
In `_types.py` we find this line:

    RequestData = dict

which affects the type hint in `httpx.Request`:

    class Request:
        def __init__(
            # ... some params omitted ...
            data: RequestData = None,
            # ... some params omitted ...
        ):

That means that forming a request with a string, or sending directly with one of the convenience functions, gives a type hint warning in a decent IDE:

    already_stringified_data = json.dumps({"foo": "bar"})
    client.send(Request(url, "POST", data=already_stringified_data))  # Warning: passing str instead of dict
    client.post(url, data=already_stringified_data)  # Warning: passing str instead of dict

Looking at the implementation of `Request` (and `PreparedRequest`), it appears that using a string like this is valid usage. And it would be a pity, if you've already encoded your JSON for some other reason, to pass the original `dict` to httpx just so it can redundantly encode it.

If the above is a valid supported usage, I'd say that this is a bug. As the title says, `RequestData` ought to be `Union[dict,str,bytes]`.

**Comment by adriangb** [comment]
Agreed!
